src/common/device_helpers.cuh

```diff
@@ -142,31 +142,91 @@ DEV_INLINE void AtomicOrByte(unsigned int* __restrict__ buffer, size_t ibyte, un
+    }
+  }
+
+  return left ? items_begin - items : items + n - items_begin;
+}
+
+}
+
 /*!
  * \brief Find the strict upper bound for an element in a sorted array
  *  using binary search.
- * \param cuts pointer to the first element of the sorted array
+ * \param items pointer to the first element of the sorted array
  * \param n length of the sorted array
  * \param v value for which to find the upper bound
- * \return the smallest index i such that v < cuts[i], or n if v is greater or equal
- *  than all elements of the array
+ * \param comp determines how the items are sorted ascending/descending order - should conform
+ *        to ordering semantics
+ * \return the smallest index i that has a value > v, or n if none is larger when sorted ascendingly
+ *         or, an index i with a value < v, or 0 if none is smaller when sorted descendingly
 */
-template <typename T>
-DEV_INLINE int UpperBound(const T* __restrict__ cuts, int n, T v) {
-  if (n == 0)           { return 0; }
-  if (cuts[n - 1] <= v) { return n; }
-  if (cuts[0] > v)      { return 0; }
-
-  int left = 0, right = n - 1;
-  while (right - left > 1) {
-    int middle = left + (right - left) / 2;
-    if (cuts[middle] > v) {
-      right = middle;
-    } else {
-      left = middle;
-    }
+// Preserve existing default behavior of upper bound
+template <typename T, typename Comp = thrust::less<T>>
+XGBOOST_DEVICE __forceinline__ uint32_t UpperBound(const T *__restrict__ items,
+                                                   uint32_t n,
+                                                   T v,
+                                                   const Comp &comp = Comp()) {
+  if (std::is_same<Comp, thrust::less<T>>::value ||
+      std::is_same<Comp, thrust::greater<T>>::value) {
+    return n - internal::CountNumItemsImpl(false, items, n, v, comp);
+  } else {
+    static_assert(std::is_same<Comp, thrust::less<T>>::value ||
+                  std::is_same<Comp, thrust::greater<T>>::value,
+                  "Invalid comparator used in Upperbound - can only be thrust::greater/less");
+    return std::numeric_limits<uint32_t>::max(); // Simply to quiesce the compiler
+  }
+}
+
+/*!
+ * \brief Find the strict lower bound for an element in a sorted array
+ *  using binary search.
+ * \param items pointer to the first element of the sorted array
+ * \param n length of the sorted array
+ * \param v value for which to find the upper bound
+ * \param comp determines how the items are sorted ascending/descending order - should conform
+ *        to ordering semantics
+ * \return the smallest index i that has a value >= v, or n if none is larger
+ *         when sorted ascendingly
+ *         or, an index i with a value <= v, or 0 if none is smaller when sorted descendingly
+*/
+// Preserve existing default behavior of upper bound
+template <typename T, typename Comp = thrust::less<T>>
+XGBOOST_DEVICE __forceinline__ uint32_t LowerBound(const T *__restrict__ items,
```

Comment:
It's a bit unusual to use a comparison operator in this way to define the sort direction of the input. For this kind of code I'm inclined to stay as close to the STL/thrust as possible so it's immediately obvious whats happening.

A more 'STL-like' way to achieve this could be to pass a reverse iterator to the standard lower bound function.

Reply:
Nevermind I just saw that they have a comparison operator in thrust :)

Reply:
i modeled this api to be similar to the std/thrust's lower and upper bound, which also takes in a comparator. the only different is that this api returns the position as opposed to an iterator (as that is how it was used before in xgboost), where an item could be slotted without violating the order of the items determined by the comparator (which determines which one comes before the other).